### PR TITLE
[lang] Update proc-macro2, quote, syn to 1.0

### LINKED ISSUE #166: Update to proc_macro2, quote and syn 1.0
With the recent release of `proc_macro2`, `quote` and `syn` version 1.0 we want to update all our references to those crates within ink! to version 1.0.
This requires touching of `ink_lang`, `ink_abi_derive`.

The `syn` update might require some significant updates to both `ink_lang` and `ink_abi_derive`.

### PR DESCRIPTION
**Pull request by koushiro**
Signed-off-by: koushiro <koushiro.cqx@gmail.com>

Close #166

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,6 @@ Cargo.lock
 
 # Ignore history files.
 **/.history/**
+
+# Ignore Intellij IDEA files
+/.idea
```
Comment (Robbepop) [inline .gitignore]: Please use your local git ignore to store information like this. You are also allowed to remove `/design`, `.history` from here.

File: cli/src/cmd/build.rs [hunk 2]
```diff
@@ -242,9 +242,6 @@ mod tests {
 
     #[cfg(feature = "test-ci-only")]
     #[test]
-    // FIXME: https://github.com/paritytech/ink/issues/202
-    // currently fails on CI because of global RUSTFLAGS overriding required `--import-memory`
-    #[ignore]
```
Comment (Robbepop) [inline cli/src/cmd/build.rs]: Why remove this comment? Is it outdated?

File: lang/src/ast.rs [hunk 3]
```diff
@@ -352,36 +358,39 @@ impl FnArg {
...
 
 #[derive(Debug)]
-pub struct ArgCaptured {
-    pub pat: syn::Pat,
+pub struct PatType {
+    pub attrs: Vec<syn::Attribute>,
+    pub pat: Box<syn::Pat>,
     pub colon_token: Token![:],
-    pub ty: syn::Type,
+    pub ty: Box<syn::Type>,
```
Comment (Robbepop) [inline lang/src/ast.rs]: Why is `Box` required here?

File: lang/src/parser.rs [hunk 4]
```diff
@@ -113,10 +117,7 @@ impl ast::ItemEnvTypesMeta {
             .segments
             .first()
             .expect("paths have at least one segment");
-        if let Some(colon) = first_segment.punct() {
-            return Err(syn::Error::new(colon.spans[0], "expected meta value"))
-        }
```
Comment (Robbepop) [inline lang/src/parser.rs]: Why remove this check?

File: .gitignore (hunk 1 above)
Comment (koushiro) [reply .gitignore]: okay

File: cli/src/cmd/build.rs (hunk 2 above)
Comment (koushiro) [reply cli/src/cmd/build.rs]: yes, fixed in #216

File: lang/src/parser.rs (hunk 4 above)
Comment (koushiro) [reply lang/src/parser.rs]: Fixed

File: lang/src/ast.rs (hunk 3 above)
Comment (koushiro) [reply lang/src/ast.rs]: `PatType` and `Pat` are large enum variants. 
Just like [PatType](https://docs.rs/syn/1.0.8/syn/struct.PatType.html) and [Pat](https://docs.rs/syn/1.0.8/syn/enum.Pat.html) in syn.

Comment (Robbepop) [reply lang/src/ast.rs]: okay sounds fair

**Pull request merged**